sky/skylet/providers/oci/node_provider.py

```diff
@@ -0,0 +1,440 @@
+                break
+            except Exception as e:
+                retry_count = retry_count + 1
+                wait_seconds = oci_conf.RETRY_INTERVAL_BASE_SECONDS * retry_count
+                logger.warn(f"! Not ready yet, wait {wait_seconds} seconds & retry!")
+                logger.warn(f"! Exception message is {str(e)}")
+                time.sleep(wait_seconds)
+    # end set_node_tags(...)
+
+
+    @synchronized
+    def terminate_node(self, node_id):
+        # """Terminates the specified node."""
+        logger.info(f"* terminate_node {node_id}...")  
+        node = self._get_node(node_id)
+    
+        logger.debug(f"* sky_spot_flag: {node['tags']['sky_spot_flag']}")
+        preemptibleFlag = True if node and (
+            str(node['tags']['sky_spot_flag']) == 'true') else False
+        
+        stopping_insts = []
+        if self.cache_stopped_nodes and not preemptibleFlag:
+            logger.info(
+                f"* Stopping instance {node_id}"
+                "(to fully terminate instead, "
+                "set `cache_stopped_nodes: False` "
+                "under `provider` in the cluster configuration)"
+            )
+            instance_action_response = oci_conf.core_client.instance_action(
+                instance_id = node_id, action="STOP")
+            logger.info(f"* Stopping the instance {instance_action_response.data.id}")
+            get_instance_response = oci_conf.core_client.get_instance(instance_id = node_id)
+            logger.debug(get_instance_response.data)
+            stopping_insts.append(node_id)
+        else:
+            terminate_instance_response = oci_conf.core_client.terminate_instance(node_id)
+            logger.debug(terminate_instance_response.data)
+            if node_id in self.cached_nodes:
+                del self.cached_nodes[node_id]
+
+        for stopping_node_id in stopping_insts:
+            # Better to treat stopping as stopped?
+            # oci.wait_until(oci_conf.core_client, get_instance_response, 'lifecycle_state', 'STOPPED')
+            if stopping_node_id in self.cached_nodes:
+                self.cached_nodes[stopping_node_id]["status"] = "STOPPED"
+
+        logger.info(f"* Teminated {node_id} w/ sky_spot_flag: {preemptibleFlag}.")
+    # end terminate_node(...)
+
+
+    def _get_node(self, node_id):
+        self._get_filtered_nodes({}) # All except for those terminated.
+        return self.cached_nodes.get(node_id, None)
+
+
+    def _get_cached_node(self, node_id):
+        if node_id in self.cached_nodes:
+            return self.cached_nodes[node_id]
+        return self._get_node(node_id = node_id)
+    
+
+    def stopped_nodes(self, tag_filters):
+        """Return a list of stopped node ids filtered by the specified tags dict."""
+        nodes = self._get_filtered_nodes(tag_filters = tag_filters)
+        return [k for k, v in nodes.items() if v["status"] == "STOPPED"]
```

Comment:
Do we need to include the `STOPPING` nodes here, assuming the `STOPPING` instance will finally be transferred to `STOPPED` status?

Reply:
BTW, a completed catalog is neccessary, but not easy because different region, and different availability domain (zone) in a region, have different resource available. We need to make sure the shapes listed in the catalog file are avaiable ones for the specified region+zone. We have plan to do that, probably with the help of product teams. Let's prioritize it after this PR

Reply:
> Do we need to include the `STOPPING` nodes here, assuming the `STOPPING` instance will finally be transferred to `STOPPED` status?

We'd better not include "STOPPING" here, because the "STOPPED" nodes are usable/startable while creating new nodes, but not for the "STOPPING" ones . If we include "STOPPING" nodes here, we may need more sophisticated logic, for example, wait the 'STOPPING' ones to be 'STOPPED' before trying to reuse them. Maybe i remark this as enhancement points and add it to the memo note at this stage.

Reply:
> BTW, a completed catalog is neccessary, but not easy because different region, and different availability domain (zone) in a region, have different resource available. We need to make sure the shapes listed in the catalog file are avaiable ones for the specified region+zone. We have plan to do that, probably with the help of product teams. Let's prioritize it after this PR

Leaving the complete catalog to the future sounds good to me! Let's have the MVP first.

> We'd better not include "STOPPING" here, because the "STOPPED" nodes are usable/startable while creating new nodes, but not for the "STOPPING" ones . If we include "STOPPING" nodes here, we may need more sophisticated logic, for example, wait the 'STOPPING' ones to be 'STOPPED' before trying to reuse them. Maybe i remark this as enhancement points and add it to the memo note at this stage.

We may need to have this `STOPPING` to avoid resource leakage. By resource leakage, it can happens in the following case:
1. `sky launch -c min --cloud oci`
2. `sky stop min`
3. During the cluster `min` is stopping, the user run `sky start min`
4. Since the original cluster is in `STOPPING` status, the `create_node` function will ignore the original cluster and start another one, which will cause the original cluster not being tracked by the SkyPilot and leaked.

We may have one of the following two options to resolve the problem:
1. Have the `STOPPING` cluster here, and handle the `STOPPING` status in the `create_node` when reusing the existing cluster. Here is how we do it for AWS: https://github.com/skypilot-org/skypilot/blob/8d3fce2a166a1d83b741cdabdba152b0e2352623/sky/skylet/providers/aws/node_provider.py#L357-L359
2. In `terminate_node` we wait the instance to be `STOPPED` before we return. Here is how we do it for GCP: https://github.com/skypilot-org/skypilot/blob/8d3fce2a166a1d83b741cdabdba152b0e2352623/sky/skylet/providers/gcp/node_provider.py#L310-L328

Reply:
Actuall i already had the logic 2 "In terminate_node we wait the instance to be STOPPED before we return", but have it commented out because it needs cost seconds to wait a instance from STOPPING to STOPPED.  Let me uncomment it.